test/test_torch.py

```diff
@@ -4644,6 +4607,138 @@ def ref_index_select(src, dim, idx):
+                if accumulate:
+                    idx = make_idx(src_size, high=dst.numel())
+                else:
+                    idx = torch.randperm(dst.numel(), dtype=torch.int64, device=device)[:src_size[0]]
+                if not idx_contig:
+                    idx = torch.repeat_interleave(idx, 2, dim=-1)[..., ::2]
+                if idx_reshape:
+                    idx = idx.reshape(2, 2)
+                out = torch.put(dst, idx, src, accumulate)
+                # out-place
+                reference = ref_put(dst, idx, src, accumulate)
+                self.assertEqual(out, reference)
+
+                # in-place
+                dst.put_(idx, src, accumulate)
+                self.assertEqual(dst, reference)
+
+
+        # Create the 8 possible combinations of scalar sizes for target / index / source
+        scalars = ((make_arg(size_t),
+                    make_idx(size_i, high=1),
+                    make_arg(size_s))
+                   for size_t, size_i, size_s in product([(), (1,)], repeat=3))
+        for (dest, idx, source), accumulate in product(scalars, [True, False]):
+            dest_init = dest.clone()
+            # out-place
+            out = torch.put(dest, idx, source, accumulate=accumulate)
+            # in-place
+            dest1 = dest.clone()
+            dest1.put_(idx, source, accumulate=accumulate)
+            for d in [out, dest1]:
+                if accumulate:
+                    self.assertEqual(d.item(), (dest_init + source).item())
+                else:
+                    self.assertEqual(d.item(), source.item())
+
+        # Empty case
+        dest = make_arg((3, 2))
+        reference = dest.clone()
+        idx = make_idx((0,), high=1)
+        source = make_arg((0,))
+        for accumulate in [True, False]:
+            out = torch.put(dest, idx, source, accumulate=accumulate)
+            self.assertEqual(out, reference)
+            dest.put_(idx, source, accumulate=accumulate)
+            self.assertEqual(dest, reference)
+
+    @dtypes(*torch.testing.get_all_dtypes(include_bfloat16=False,
+                                          include_half=False,
+                                          include_bool=False))
+    def test_put_accumulate(self, device, dtype):
+        # Test for parallel adds with accumulate == True
+        make_arg = partial(make_tensor, low=-4, high=5, device=device, dtype=dtype)
+        # Dump everything into the 0-th position
+        make_idx = partial(torch.zeros, device=device, dtype=torch.int64)
+
+        # Grainsize is 3000 for the for_loop to be parallized on CPU
+        for idx, source in ((make_idx(size), make_arg(size)) for size in ((200,), (3002,))):
+            dest = make_arg((1,))
+            dest2 = dest.clone()
+            dest.put_(idx, source, accumulate=True)
+            self.assertEqual(dest, dest2 + source.sum())
+
+    @dtypes(torch.half, torch.bfloat16)
+    def test_put_accumulate_halfs(self, device, dtype):
```

Comment:
why do you need a separate test for half and bfloat16? It looks like half shouldn't overflow even on you regular test, and you can set higher dtype-dependent tolerance to handle larger errors.

Reply:
When done tweaking `rtol` and `atol` the tests are quite unstable. An example with `bfloat`:
`The greatest difference was 96.0 (1504.0 vs. 1408.0), which occurred at index `.
I suggest mergint the two tests, but leaving the same ranges and vector sizes as are being tested now.